From e01979d176b077956311116c1b43e5146c67ebd7 Mon Sep 17 00:00:00 2001
From: Zeng Zhaoming <b32542@freescale.com>
Date: Wed, 11 May 2011 08:39:56 +0800
Subject: [PATCH 961/998] ENGR00143384 Fec: fix get wrong carrier status when ethx down

using ioctl() to get interface carrier status, when ethx down,
will get wrong status, accroding to linux net subsystem, this case
should return EINVAL since phy not start and no carrier status can
get.

Signed-off-by: Zeng Zhaoming <b32542@freescale.com>
---
 drivers/net/fec.c |   10 +++++++++-
 1 files changed, 9 insertions(+), 1 deletions(-)

diff --git a/drivers/net/fec.c b/drivers/net/fec.c
index b85f196..49a9890 100644
--- a/drivers/net/fec.c
+++ b/drivers/net/fec.c
@@ -914,11 +914,19 @@ static void fec_enet_get_drvinfo(struct net_device *dev,
 	strcpy(info->bus_info, dev_name(&dev->dev));
 }
 
+static int fec_enet_get_link(struct net_device *dev)
+{
+	if (netif_running(dev))
+		return netif_carrier_ok(dev) ? 1 : 0;
+	else
+		return -EINVAL;
+}
+
 static struct ethtool_ops fec_enet_ethtool_ops = {
 	.get_settings		= fec_enet_get_settings,
 	.set_settings		= fec_enet_set_settings,
 	.get_drvinfo		= fec_enet_get_drvinfo,
-	.get_link		= ethtool_op_get_link,
+	.get_link		= fec_enet_get_link,
 };
 
 static int fec_enet_ioctl(struct net_device *dev, struct ifreq *rq, int cmd)
-- 
1.7.4.1

